src/_pytest/pathlib.py

```diff
@@ -305,11 +305,13 @@ def make_numbered_dir_with_cleanup(
             e = exc
         else:
             consider_lock_dead_if_created_before = p.stat().st_mtime - lock_timeout
-            cleanup_numbered_dir(
-                root=root,
-                prefix=prefix,
-                keep=keep,
-                consider_lock_dead_if_created_before=consider_lock_dead_if_created_before,
+            # Register a cleanup for program exit
+            atexit.register(
```

Comment:
This might still fail because some pytest process will finish before others so they might still step in each others toes, I guess?

Reply:
Do you mean one of the race condition cases mentioned in maybe_delete_a_numbered_dir()? I thought try_cleanup() invoked from cleanup_numbered_dir() is preventing from such case or at least we have some guarantees - based on comments from maybe_delete_a_numbered_dir().